File: absolute-zero/examples/javascript/nop.js
#!/usr/bin/env node
/**
 * JavaScript CNO (Code with No Output) - Absolute Zero Example
 *
 * This program executes successfully without producing any output.
 * Compatible with both Node.js and browser environments.
 * Demonstrates the minimal JavaScript program that does nothing observably.
 *
 * INTERPRETER BEHAVIOR (Node.js):
 * - V8 engine initialization and JIT compilation
 * - Event loop creation (libuv)
 * - Global object setup (process, Buffer, require, etc.)
 * - Runtime overhead: ~30-60ms for Node.js startup
 *
 * INTERPRETER BEHAVIOR (Browser):
 * - JavaScript engine initialization (V8, SpiderMonkey, JavaScriptCore)
 * - DOM not accessed, so no layout/rendering triggered
 * - Window object already exists
 * - Runtime overhead: <1ms (engine already running)
 *
 * RUNTIME OVERHEAD (Node.js):
 * - V8 heap initialization (~4-10 MB baseline)
 * - Built-in modules loaded into require cache
 * - Event loop thread spawned
 * - Process object populated with env, argv, etc.
 * - Approximate cost: 20-30 MB memory footprint minimum
 *
 * RUNTIME OVERHEAD (Browser):
 * - Minimal - script parsed and executed in existing context
 * - No new memory allocation beyond script object
 * - Approximate cost: <100 KB for script parse/compile
 *
 * SIDE EFFECTS (Node.js):
 * - process.exit(0) called implicitly when event loop empty
 * - process.exitCode set to 0 by default
 * - File descriptors 0, 1, 2 inherited (stdin, stdout, stderr)
 * - process.hrtime() affected (timing visible)
 * - No files created (unlike Python .pyc)
 *
 * SIDE EFFECTS (Browser):
 * - Script element execution state changes
 * - Performance timeline entries created
 * - JavaScript task added to event loop history
 * - Memory profiler sees script allocation
 *
 * VERIFICATION:
 * - No console output
 * - No stdout/stderr output (Node.js)
 * - Exit code 0 (Node.js)
 * - No exceptions raised
 * - No DOM modifications (browser)
 *
 * EXECUTION:
 *     # Node.js
 *     node nop.js
 *     echo $?  # Should output 0
 *
 *     # Browser
 *     <script src="nop.js"></script>
 *
 * LANGUAGE SEMANTICS:
 * - Empty file is valid JavaScript program
 * - undefined is implicit return value
 * - Comments-only file is valid
 * - Strict mode has no effect on empty program
 */

'use strict';

// Explicit no-operation (uncomment to demonstrate):
// void 0;

// Alternative explicit no-ops in JavaScript:
// undefined;
// (function(){})();
// (() => {})();
// {};
// [];
// null;

// Note: All of the above are valid statements that do nothing observable
// The semicolons are optional due to ASI (Automatic Semicolon Insertion)

File: absolute-zero/examples/fsharp/Nop.fs
(*
 * Certified Null Operation in F#
 *
 * A functional program that does absolutely nothing at the application level.
 * Returns exit code 0 without any observable side effects.
 *
 * Properties:
 * - Terminates immediately
 * - No I/O operations
 * - No side effects (pure functional CNO)
 * - Exit code 0
 *
 * Compile: fsharpc Nop.fs
 * Run: ./Nop.exe (or mono Nop.exe)
 *)

[<EntryPoint>]
let main argv =
    // The minimal CNO in F#
    // Returns 0 to indicate successful termination
    // No effectful computation occurs
    0

(*
 * Verification notes:
 * - F# runtime initializes (.NET CLR)
 * - Functional programming emphasizes referential transparency
 * - This program has no side effects - a "pure" CNO
 * - The return value 0 is the only output
 *
 * In functional programming terms, this is:
 * main : string[] -> int
 * main = λargs. 0
 *
 * The function ignores its input and returns zero,
 * performing no effectful computation in the process.
 *)
